File: controllers/salmon.py
import datetime
import re
import urllib2
import urlparse

from BeautifulSoup import BeautifulSoup
import magicsig
import salmoning
import tornado.escape
import tornado.web

from base import BaseHandler
from logic import content_remote
from logic import smtp
from logic import socialize
from logic import spam
from logic import url_factory
from logic import users

class SalmonHandler(BaseHandler):
  class MockKeyRetriever(magicsig.KeyRetriever):
    def __init__(self, handler, local_user):
      self.handler = handler
      self.local_user = local_user
      self.signer_uri = None

    def LookupPublicKey(self, signer_uri=None):
      self.signer_uri = signer_uri
      user_remote = self.handler.models.users_remote.get(
          local_username=self.local_user.username, profile_url=signer_uri)[0]

      if not user_remote:
        # get host-meta first
        lrdd_link = users.get_lrdd_link(signer_uri)

        if not lrdd_link:
          raise tornado.web.HTTPError(400)

        # get webfinger
        webfinger_doc = users.get_webfinger(lrdd_link, signer_uri)
        magic_key = webfinger_doc.find('link',
            rel='magic-public-key')['href'].replace(
            'data:application/magic-public-key,', '')

        if not magic_key:
          raise tornado.web.HTTPError(400)

        new_user = self.handler.models.users_remote()
        new_user.local_username = self.local_user.username
        new_user.profile_url = signer_uri
        new_user.magic_key = magic_key
        new_user.save()

        return (str(magic_key))
      else:
        return (str(user_remote.magic_key))

  def post(self):
    acct = self.get_argument("q").split(':')

    # remove acct: if it's there
    if len(acct) > 1:
      acct[0] = acct[1].split('@')[0]

    acct = acct[0].split('@')[0]

    self.display["acct"] = acct
    user = self.models.users.get(username=acct)[0]

    if not user:
      raise tornado.web.HTTPError(404)

    self.display["user"] = user

    salmonizer = salmoning.SalmonProtocol()
    salmonizer.key_retriever = self.MockKeyRetriever(handler=self,
        local_user=user)
    env = salmonizer.ParseSalmon(self.request.body,
        'application/magic-envelope+xml')
    self.handle_activity_verb(env, salmonizer.key_retriever.signer_uri)

  # tornado & python rock
  def check_xsrf_cookie(self):
    pass


  def handle_activity_verb(self, env, signer_uri):
    salmon_doc = BeautifulSoup(env)
    activity_verb = salmon_doc.find(re.compile('.+:verb$')).string
    user_remote = self.models.users_remote.get(
        local_username=self.display["user"].username,
        profile_url=signer_uri)[0]

    if (activity_verb == 'http://activitystrea.ms/schema/1.0/follow'):
      user = users.get_remote_user_info(self, signer_uri,
          user_remote.local_username)
      user.follower = 1
      user.save()
      smtp.follow(self, user.username, self.display["user"].email,
          user.profile_url)
    elif (activity_verb == 'http://ostatus.org/schema/1.0/unfollow' or
        activity_verb == 'http://activitystrea.ms/schema/1.0/stop-following'):
      user_remote.follower = 0
      user_remote.save()
    elif (activity_verb == 'http://activitystrea.ms/schema/1.0/favorite'):
      activity_object = salmon_doc.find(re.compile('.+:object$'))
      atom_id = activity_object.find('atom:id').string
      local_url = atom_id.split(':')[2]
      content_url = url_factory.load_basic_parameters(self, url=local_url)

      already_favorited = self.models.content_remote.get(
          to_username=self.display["user"].username,
          from_user=signer_uri,
          type='favorite',
          local_content_name=content_url['name'])[0]
      if already_favorited:
        return

      content = self.models.content.get(username=self.display["user"].username,
                                        name=content_url['name'])[0]
      if not content:
        raise tornado.web.HTTPError(400)
      content.favorites += 1
      content.save()

      favorite_record = self.models.content_remote()
      favorite_record.to_username = self.display["user"].username
      favorite_record.from_user = signer_uri
      favorite_record.username = user_remote.username
      favorite_record.date_created = datetime.datetime.strptime(
          salmon_doc.find('atom:updated').string[:-6], '%Y-%m-%dT%H:%M:%S')
      favorite_record.type = 'favorite'
      favorite_record.local_content_name = content.name
      favorite_record.save()
    elif (activity_verb == 'http://activitystrea.ms/schema/1.0/unfavorite' or
        activity_verb == 'http://ostatus.org/schema/1.0/unfavorite'):
      # TODO no activity object at least with ostatus??
      pass
    elif (activity_verb == 'http://activitystrea.ms/schema/1.0/share'):
      # TODO
      pass
    elif (activity_verb == 'http://activitystrea.ms/schema/1.0/post'):
      atom_content = salmon_doc.find('atom:content').string
      sanitized_atom_content = content_remote.sanitize(
          tornado.escape.xhtml_unescape(atom_content))

      existing_content = self.models.content_remote.get(
          to_username=self.display["user"].username,
          from_user=signer_uri,
          view=sanitized_atom_content)[0]

      thread = salmon_doc.find('thr:in-reply-to')
      ref = ''
      if thread:
        try:
          local_url = thread['ref'].split(':')[2]
          content_url = url_factory.load_basic_parameters(self, url=local_url)
          content = self.models.content.get(
              username=self.display["user"].username,
              name=content_url['name'])[0]
          if not content:
            raise tornado.web.HTTPError(400)
          ref = content_url['name']
          content.comments_count += 1
          content.comments_updated = datetime.datetime.utcnow()
          content.save()
        except Exception as ex:
          import logging
          logging.error("something wrong with thread")
          logging.error(ex)

      replies = salmon_doc.find('thr:replies')
      comments_count = 0
      comments_updated = None
      if replies:
        if replies.has_key('count'):
          comments_count = int(replies['count'])
        if replies.has_key('updated'):
          comments_updated = replies['updated']
        comments_response = urllib2.urlopen(replies['href'])
        content_remote.parse_feed(self.models, user_remote,
            comments_response.read(), remote_comments=True)

      mentioned = salmon_doc.findAll('atom:link', rel='mentioned')
      if not mentioned:
        mentioned = salmon_doc.findAll('atom:link', rel='ostatus:attention')

      this_user_mentioned = False
      if mentioned:
        this_user_url = self.nav_url(host=True,
            username=self.display["user"].username)
        for mentions in mentioned:
          if mentions['href'] == this_user_url:
            # hey, you've been mentioned. cool.
            this_user_mentioned = True
            break

      is_spam = spam.guess(atom_content,
          self.application.settings["private_path"],
          self.display["user"].username)

      if existing_content:
        # possible that it's picked up via feed, before we get the salmon call
        post_remote = existing_content
      else:
        post_remote = self.models.content_remote()
      post_remote.to_username = self.display["user"].username
      post_remote.from_user = signer_uri
      post_remote.username = user_remote.username
      post_remote.avatar = user_remote.avatar
      post_remote.date_created = datetime.datetime.strptime(
          salmon_doc.find('atom:published').string[:-6], '%Y-%m-%dT%H:%M:%S')
      post_remote.date_updated = datetime.datetime.strptime(
          salmon_doc.find('atom:updated').string[:-6], '%Y-%m-%dT%H:%M:%S')
      post_remote.comments_count = comments_count
      if comments_updated:
        post_remote.comments_updated = datetime.datetime.strptime(
            comments_updated[:-6], '%Y-%m-%dT%H:%M:%S')
      else:
        post_remote.comments_updated = None
      if is_spam:
        post_remote.is_spam = True
      else:
        spam.train_ham(atom_content, self.application.settings["private_path"],
            self.display["user"].username)
      post_remote.type = 'comment' if ref or (existing_content and
          existing_content.type == 'comment') else 'post'
      post_remote.title = salmon_doc.find('atom:title').string
      post_remote.post_id = salmon_doc.find('atom:id').string
      post_remote.link = salmon_doc.find('atom:link', rel='alternate')['href']
      post_remote.local_content_name = content.name if ref else ''
      post_remote.view = sanitized_atom_content
      post_remote.save()

      if ref:
        socialize.socialize(self, content)

      if ref:
        smtp.comment(self, post_remote.username, post_remote.from_user,
            self.display["user"].email, self.content_url(content, host=True),
            sanitized_atom_content)
      elif this_user_mentioned:
        smtp.comment(self, post_remote.username, post_remote.from_user,
            self.display["user"].email, post_remote.link,
            sanitized_atom_content, this_user_mentioned=True)

File: server/api/social_butterfly/salmon.js
import { buildUrl, contentUrl, parseContentUrl, profileUrl } from '../../../shared/util/url_factory';
import cheerio from 'cheerio';
import fetch from 'node-fetch';
import { follow as emailFollow } from './email';
import { getUserRemoteInfo } from './discover_user';
import magic from 'magic-signatures';
import { mention as emailMention } from './email';
import models from '../../data/models';
import { parseUsernameFromAccount } from './discover_user';
import React, { createElement as RcE, PureComponent } from 'react';
import { renderToString } from 'react-dom/server';
import { sanitizeHTML } from '../../util/crawler';
import socialize from './socialize';

export async function follow(req, contentOwner, salmonUrl, isFollow) {
  const action = isFollow ? 'follow' : 'unfollow';
  send(
    req,
    contentOwner,
    salmonUrl,
    <Salmon req={req} action={action} title={action} atomContent={action} verb={action} />
  );
}

export async function favorite(req, contentOwner, contentRemote, salmonUrl, isFavorite) {
  const action = isFavorite ? 'favorite' : 'unfavorite';
  const activityObject = RcE(
    'activity:object',
    {},
    RcE('activity:object-type', {}, 'http://activitystrea.ms/schema/1.0/note'),
    <id>{contentRemote.post_id}</id>,
    <title />,
    <content type="html" />
  );
  send(
    req,
    contentOwner,
    salmonUrl,
    <Salmon
      req={req}
      activityObject={activityObject}
      action={action}
      title={action}
      atomContent={action}
      verb={action}
    />
  );
}

export async function reply(req, contentOwner, content, salmonUrl, mentionedRemoteUsers) {
  const contentUrlWithHost = contentUrl(content, req);
  const contentUrlWithoutHost = contentUrl(content);
  const objectType = content.section === 'comments' ? 'comment' : 'note';
  const repliesUrl = buildUrl({ pathname: '/api/social/comments', searchParams: { url: contentUrlWithoutHost } });
  const activityObject = (
    <>
      {RcE('activity:object-type', {}, `http://activitystrea.ms/schema/1.0/${objectType}`)}
      <link href={contentUrlWithHost} rel="alternate" type="text/html" />
      {mentionedRemoteUsers.map(mentionedRemoteUser => (
        <>
          <link href={mentionedRemoteUser.profile_url} rel="ostatus:attention" />
          <link href={mentionedRemoteUser.profile_url} rel="mentioned" />
        </>
      ))}

      {/* see endpoint_with_apollo for refXXX transform */}
      {content.thread ? RcE('thr:in-reply-to', { refXXX: content.thread }) : null}
      {content.comments_count
        ? RcE('thr:replies', {
            type: 'application/atom+xml',
            href: repliesUrl,
            count: content.comments_count,
            updated: new Date(content.comments_updated).toISOString(),
          })
        : null}
    </>
  );

  send(
    req,
    contentOwner,
    salmonUrl,
    <Salmon
      req={req}
      activityObject={activityObject}
      id={contentUrlWithHost}
      action={'post'}
      content={content}
      title={content.title}
      atomContent={content.view}
      verb={'post'}
    />
  );
}

async function send(req, contentOwner, salmonUrl, tree) {
  if (!salmonUrl) {
    return;
  }

  let renderedTree = renderToString(tree);
  // XXX(mime): in the feeds I have some attributes that are `ref`. However, ref isn't allowed in React,
  // so in the DOM they are `refXXX`. Return them to normal here, sigh.
  renderedTree = renderedTree.replace(/refXXX="([^"]+)"/g, 'ref="$1"');

  const body = magic.sign({ data: renderedTree, data_type: 'application/atom+xml' }, contentOwner.private_key);
  try {
    await fetch(salmonUrl, {
      method: 'POST',
      body,
      headers: {
        'Content-Type': 'application/magic-envelope+xml',
      },
    });
  } catch (ex) {
    // Not a big deal if this fails.
    // TODO(mime): add logging later.
  }
}

class Salmon extends PureComponent {
  render() {
    const { req, activityObject, atomContent, content, id, title, verb } = this.props;
    const namespaces = {
      xmlLang: 'en-US',
      xmlns: 'http://www.w3.org/2005/Atom',
      'xmlns:activity': 'http://activitystrea.ms/spec/1.0/',
      'xmlns:thr': 'http://purl.org/syndication/thread/1.0',
    };
    const tagDate = new Date().toISOString().slice(0, 10);

    return (
      <entry {...namespaces}>
        <title type="html">{title}</title>
        <id>{`tag:${req.get('host')},${tagDate}:${id || new Date().toISOString()}`}</id>
        <content type="html" dangerouslySetInnerHTML={{ __html: atomContent }} />
        <published>{new Date(content?.createdAt || new Date()).toISOString()}</published>
        <updated>{new Date(content?.updatedAt || new Date()).toISOString()}</updated>
        {RcE('activity:verb', {}, `http://activitystrea.ms/schema/1.0/${verb}`)}
        {activityObject}
      </entry>
    );
  }
}

export default async (req, res) => {
  // TODO(mime): make sure xsrf check here is disabled.
  if (!req.query.q) {
    return res.sendStatus(400);
  }
  const username = parseUsernameFromAccount(req.query.q);
  const user = await models.User.findOne({ where: { username } });
  if (!user) {
    return res.sendStatus(404);
  }

  const magicJsonBody = magic.fromXML(req.body);
  const websiteUrl = magicJsonBody.author.uri;
  let userRemote = await models.User.findOne({ where: { local_username: user.username, profile_url: websiteUrl } });
  if (!userRemote) {
    const userRemoteInfo = await getUserRemoteInfo(websiteUrl, user.username);
    userRemote = await models.User_Remote.create(userRemoteInfo, { validate: true });
  }

  const env = magic.verify(magicJsonBody, userRemote.magic_key);
  const $ = cheerio.load(env);
  const verb = $('activity:verb').text();

  switch (verb) {
    case 'http://activitystrea.ms/schema/1.0/follow':
      receiveFollow(req, user, userRemote, true);
      break;
    case 'http://ostatus.org/schema/1.0/unfollow':
    case 'http://activitystrea.ms/schema/1.0/stop-following':
      receiveFollow(req, user, userRemote, false);
      break;
    case 'http://activitystrea.ms/schema/1.0/favorite':
      receiveFavorite(res, $, websiteUrl, userRemote, true);
      break;
    case 'http://activitystrea.ms/schema/1.0/unfavorite':
    case 'http://ostatus.org/schema/1.0/unfavorite':
      receiveFavorite(res, $, websiteUrl, userRemote, false);
      break;
    case 'http://activitystrea.ms/schema/1.0/share':
      // Nothing here yet.
      break;
    case 'http://activitystrea.ms/schema/1.0/post':
      receivePostOrComment(req, res, $, user, userRemote, websiteUrl);
      break;
    default:
      break;
  }

  res.sendStatus(204);
};

async function receiveFollow(req, user, userRemote, isFollow) {
  if (isFollow) {
    await models.User_Remote.update({ follower: true }, { id: userRemote.id });
    emailFollow(req, user.username, user.email, userRemote.profile_url);
  } else {
    if (userRemote.following) {
      await models.User_Remote.update({ follower: false }, { id: userRemote.id });
    } else {
      await models.User_Remote.destroy({ where: { id: userRemote.id } });
    }
  }
}

async function receiveFavorite(res, $, websiteUrl, userRemote, isFavorite) {
  const atomId = $('activity:object atom:id').text();
  const localContentUrl = atomId.split(':')[2];
  const { username, name } = parseContentUrl(localContentUrl);

  const contentRemoteWhere = {
    from_user: websiteUrl,
    local_content_name: name,
    to_username: username,
    type: 'favorite',
    name,
  };

  if (!isFavorite) {
    await models.Content_Remote.destroy({ where: contentRemoteWhere });
    return;
  }

  const existingFavorite = await models.Content_Remote.findOne({ where: contentRemoteWhere });
  if (existingFavorite) {
    return;
  }

  const content = await models.Content.findOne({ where: { username, name } });
  if (!content) {
    return res.sendStatus(400);
  }
  await models.Content.update({ favorites_count: content.favorites_count + 1 }, { where: { id: content.id } });

  await models.Content_Remote.create({
    content: '',
    createdAt: new Date($('atom:updated').text() || new Date()),
    from_user: websiteUrl,
    link: '',
    local_content_name: content.name,
    post_id: '',
    title: '',
    to_username: username,
    type: 'favorite',
    username: userRemote.username,
    view: '',
  });
}

async function receivePostOrComment(req, res, $, user, userRemote, websiteUrl) {
  const atomContent = sanitizeHTML($('atom:content').text());
  const existingContentRemote = await models.Content_Remote.findOne({
    where: {
      to_username: user.username,
      from_user: websiteUrl,
      view: atomContent, // XXX(mime): gotta be a better way, old code.
    },
  });

  const contentRemote = {
    id: existingContentRemote?.id || undefined,
    avatar: userRemote.avatar,
    comments_count: parseInt($('thr:replies').attr('count')),
    comments_updated: new Date($('thr:replies').attr('updated') || new Date()),
    content: '',
    createdAt: new Date($('atom:published').text() || new Date()),
    from_user: websiteUrl,
    link: $('atom:link[rel="alternate"]').attr('href'),
    post_id: $('atom:id').text(),
    title: $('atom:title').text(),
    to_username: user.username,
    updatedAt: new Date($('atom:updated').text() || new Date()),
    username: userRemote.username,
    view: atomContent,
  };

  const thread = $('thr:in-reply-to');
  if (thread) {
    handlePost(req, res, contentRemote, user, $);
  } else {
    handleComment(req, res, contentRemote, thread);
  }

  //const replies = $('thr:replies');
  // if (replies) {
  //   // TODO(mime): these used to be known as 'remote-comment' types.
  //   // need to refactor this.
  //   const comments = await retrieveFeed(replies.attr('href'));
  //   await parseFeedAndInsertIntoDb(userRemote, comments);
  // }
}

async function handlePost(req, res, contentRemote, user, $) {
  const profile = profileUrl(user.username, req);
  const wasUserMentioned =
    $('atom:link[rel="mentioned"]').find(e => e.attr('href') === profile) ||
    $('atom:link[rel="ostatus:attention"]').find(e => e.attr('href') === profile);

  contentRemote.type = 'post';
  await models.Content_Remote.upsert(contentRemote);

  if (wasUserMentioned) {
    emailMention(
      req,
      'Remote User',
      contentRemote.link /* XXX(mime): can't be right */,
      user.email,
      contentRemote.link
    );
  }
}

async function handleComment(req, res, contentRemote, thread) {
  const localContentUrl = thread.attr('ref').split(':')[2];
  const { username, name } = parseContentUrl(localContentUrl);
  const content = await models.Content.findOne({ where: { username, name } });
  if (!content) {
    return res.sendStatus(400);
  }

  contentRemote.type = 'comment';
  contentRemote.local_content_name = name;
  await models.Content_Remote.upsert(contentRemote);

  await models.Content.update(
    { comments_count: content.comments_count + 1, comments_updated: new Date() },
    { where: { id: content.id } }
  );
  const updatedCommentedContent = await models.Content.findOne({ where: { id: content.id } });

  const contentOwner = await models.User.findOne({ where: { username } });
  await socialize(req, contentOwner, updatedCommentedContent, contentRemote, true /* isComment */);
}
